libraries/CurieBLE/src/BLEProfile.cpp

```diff
@@ -0,0 +1,513 @@
+    
+    //assert(notifyatt->type() == BLETypeCharacteristic);
+    pr_debug(LOG_MODULE_APP, "%s1", __FUNCTION__);
+    chrc->setValue(*((BLEHelper *)peripheral),(const unsigned char *)data, length);
+    return BT_GATT_ITER_CONTINUE;
+}
+
+// Only for central
+uint8_t profile_discover_process(struct bt_conn *conn,
+                                 const struct bt_gatt_attr *attr,
+                                 struct bt_gatt_discover_params *params)
+{
+    BLEPeripheralHelper* peripheral = BLECentralRole::instance()->peripheral(conn);// Find peripheral by bt_conn
+    peripheral->discover(attr);
+    return BT_GATT_ITER_STOP;
+}
+
+// Only for central
+uint8_t profile_read_rsp_process(struct bt_conn *conn, int err,
+                                 struct bt_gatt_read_params *params,
+                                 const void *data, 
+                                 uint16_t length)
+{
+    if (NULL == data)
+    {
+        return BT_GATT_ITER_STOP;
+    }
+    BLEPeripheralHelper* peripheral = BLECentralRole::instance()->peripheral(conn);// Find peripheral by bt_conn
+    BLEAttribute* readatt = peripheral->attribute(params->single.handle);
+    BLECharacteristic *chrc = (BLECharacteristic *)readatt;
+    
+    //assert(readatt->type() == BLETypeCharacteristic);
+    chrc->setValue(*((BLEHelper *)peripheral), (const unsigned char *)data, length);
+    return BT_GATT_ITER_STOP;
+}
+
+BLEProfile::BLEProfile (BLEPeripheralHelper *peripheral):
+    _attr_base(NULL),
+    _attr_index(0),
+    _attributes(NULL),
+    _num_attributes(0),
+    _sub_param(NULL),
+    _sub_param_idx(0)
+{
+    _peripheral = peripheral;
+    memset(&_discover_params, 0, sizeof(_discover_params));
+    _discover_params.end_handle = 0xFFFF;
+    _discover_params.start_handle = 0x0001;
+    _discover_params.func = profile_discover_process;
+}
+
+BLEProfile::~BLEProfile (void)
+{
+    if (this->_attributes) {
+        free(this->_attributes);
+    }
+    if (this->_attr_base)
+    {
+        free(this->_attr_base);
+    }
+}
+
+void BLEProfile::addAttribute (BLEAttribute& attribute)
+{
+    struct bt_gatt_attr *start;
+    if (NULL == _attributes) 
+    {
+        _attributes = (BLEAttribute**)malloc(BLEAttribute::numAttributes() * sizeof(BLEAttribute*));
+        memset(_attributes, 0x00, BLEAttribute::numAttributes() * sizeof(BLEAttribute*));
+    }
+    if (NULL == _attr_base)
+    {
+        _attr_base = (struct bt_gatt_attr *)malloc((BLEAttribute::numAttributes() + BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_attr));
+        memset(_attr_base, 0x00, ((BLEAttribute::numAttributes() + BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_attr)));
+        pr_info(LOG_MODULE_BLE, "_attr_base_-%p, size-%d", _attr_base, sizeof(_attr_base));
+    }
+    if (NULL == _sub_param)
+    {
+        _sub_param = (struct bt_gatt_subscribe_params *)malloc((BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_subscribe_params));
```

Comment:
I see that `_attributes` and `_attr_base` are freed in the `BLEProfile` class destructor, but I don't see anything in the library that frees `_sub_param`. Will this memory be lost after BLE Central is used?